### PR TITLE
Add kl_div and rel_entr functions

### LINKED ISSUE #16630: Adding `kl_div` and `rel_entropy` to `jax.scipy.special`
Hi, right now `jax.scipy.special` does not have a `kl_div` function, and computing this function naively in its closed form does not have the same numerical stability.
I checked the [implementation](https://github.com/scipy/scipy/blob/899f4efbcd8020b538bd97eb54353b3249af6e43/scipy/special/_convex_analysis.pxd#L13)  from `scipy.special.kl_div`, equivalent to:

```math
    \text{kldiv}(p,q)= \begin{cases}
    p\textrm{log}(p/q) -p +q , & \text{if } p \gt 0, q \gt 0\\
    q, & p = 0, q \ge 0 \\
    \infty,              & \text{otherwise}
\end{cases}
```
A possible LAX implementation:

```python
def kl_div(
    p: ArrayLike,
    q: ArrayLike,
) -> Array:
    both_gt_zero_mask = lax.bitwise_and(lax.gt(p, 0.0), lax.gt(q, 0.0))
    one_zero_mask = lax.bitwise_and(lax.eq(p, 0.0), lax.ge(q, 0.0))

    one_filler = lax.full_like(p, 1.0)
    inf_filler = lax.full_like(p, np.inf)

    safe_p = lax.select(both_gt_zero_mask, p, one_filler)
    safe_q = lax.select(both_gt_zero_mask, q, one_filler)

    log_val = lax.add(
        lax.sub(lax.mul(safe_p, lax.log(lax.div(safe_p, safe_q))), safe_p), safe_q
    )
    result = lax.select(
        both_gt_zero_mask, log_val, lax.select(one_zero_mask, q, inf_filler)
    )
    return result
```

Notice that scipy upgrades floats to double precision, so this JAX implementation only gives a result that matches scipy's up to an absolute tolerance of `1e-6`.
I was thinking of making a pull request adding this function, and perhaps `rel_entropy` function, (similarly implemented). What do you think?

@j-towns

### PR DESCRIPTION
**Pull request by SalamanderXing**
Their behavior is the same as functions in scipy.special. The only small difference is in rel_entr function, which unlike scipy.special does not take the optional parameter 'out', as JAX's arrays are immutable.
@jakevdp 

Resolves #16630

### CODE REVIEW

File: jax/_src/scipy/special.py [hunk 1]
```diff
@@ -36,159 +36,234 @@
 from jax._src.typing import Array, ArrayLike
 
 
-@_wraps(osp_special.gammaln, module='scipy.special')
+@_wraps(osp_special.gammaln, module="scipy.special")
 def gammaln(x: ArrayLike) -> Array:
-  x, = promote_args_inexact("gammaln", x)
-  return lax.lgamma(x)
+    (x,) = promote_args_inexact("gammaln", x)
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:41]: Please revert the unrelated formatting changes.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/special.py [hunk 2]
```diff
@@ -175,6 +175,52 @@ def entr(x: ArrayLike) -> Array:
                     lax.full_like(x, -np.inf),
                     lax.neg(_xlogx(x)))
 
+@_wraps(osp_special.kl_div, module="scipy.special")
+def kl_div(
+    p: ArrayLike,
+    q: ArrayLike,
+):
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:182]: Annotate the return type with `-> Array`

File: jax/_src/scipy/special.py [hunk 3]
```diff
@@ -175,6 +175,52 @@ def entr(x: ArrayLike) -> Array:
...
+    result = lax.select(
+        both_gt_zero_mask, log_val, lax.select(one_zero_mask, q, inf_filler)
+    )
+    return result
+
+@_wraps(osp_special.rel_entr, module="scipy.special")
+def rel_entr(
+    p: ArrayLike,
+    q: ArrayLike,
+):
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:208]: `-> Array` here as well

**Review by jakevdp — CHANGES_REQUESTED** [review]
Great start, thanks for working on this! Two minor comments below, plus a couple other things to address:

- please add tests for the new functions. The appropriate place is probably here: https://github.com/google/jax/blob/main/tests/lax_scipy_special_functions_test.py Since this is using `select` statements, it will be important to test gradients of the functions as well: I think you've employed the double-where trick in both cases, so I don't think you should run into stray NaNs, but it's best to check!
- please add the new functions to the docs by listing them here: https://github.com/google/jax/blob/main/docs/jax.scipy.rst#jaxscipyspecial
- Once all this is finished, please squash your changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests and feel free to reach out if you need help with doing that in git)

Thanks!

**Comment by SalamanderXing** [comment]
Hi @jakevdp . I applied your requested changes but I also accidentally squashed several commits into one. Can you help me resolve this issue?

**Comment by jakevdp** [comment]
You should be able to fix this by rebasing against the main branch. Assuming `origin` points to your repository, and `upstream` points to the main JAX repository, you can do this:
```
git checkout main
git pull upstream main
git checkout kl_div
git rebase main
git push origin +kl_div
```

**Comment by jakevdp** [comment]
Looks good! The last thing we'll need is to squash all your changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests). Feel free to reach out if you need help with doing that in git.

**Comment by SalamanderXing** [comment]
> Looks good! The last thing we'll need is to squash all your changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests). Feel free to reach out if you need help with doing that in git.

I should have done that, if that was unsuccessful please let me know how to squash them.

**Comment by jakevdp** [comment]
Thanks - it looks like the squashing did not work: this pull request consists of five commits. Once you squash the changes into a single commit, the pull request should consist of one commit.

File: jax/experimental/jax2tf/tests/jax_export_test.py [hunk 4]
```diff
@@ -13,8 +13,11 @@
 # limitations under the License.
 import contextlib
 import math
+<<<<<<< HEAD
 import functools
 import logging
+=======
+>>>>>>> 41620562d ([shape_poly] Improve error messages for shape assertions)
```
Comment (jakevdp) [inline jax/experimental/jax2tf/tests/jax_export_test.py:20]: Looks like some git conflicts were not properly resolved.

**Review by jakevdp — APPROVED** [review]
Looks good!

File: jax/_src/scipy/special.py [hunk 5]
```diff
@@ -191,6 +190,47 @@ def multigammaln(a: ArrayLike, d: ArrayLike) -> Array:
...
 
 
+@_wraps(osp_special.kl_div, module="scipy.special")
+def kl_div(
+    p: ArrayLike,
+    q: ArrayLike,
+) -> Array:
+    p, q = promote_args_inexact("kl_div", p, q)
+    both_gt_zero_mask = lax.bitwise_and(lax.gt(p, 0.0), lax.gt(q, 0.0))
+    one_zero_mask = lax.bitwise_and(lax.eq(p, 0.0), lax.ge(q, 0.0))
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:200]: This is failing with a dtype mismatch with float32 inputs and `JAX_ENABLE_X64=1`.

If you want to use `lax` functions here, the input types must match exactly. You could use `zero = _lax_const(p, 0.0)` to ensure this is the case.

File: jax/_src/scipy/special.py [hunk 6]
```diff
@@ -191,6 +190,47 @@ def multigammaln(a: ArrayLike, d: ArrayLike) -> Array:
...
+
+
+@_wraps(osp_special.rel_entr, module="scipy.special")
+def rel_entr(
+    p: ArrayLike,
+    q: ArrayLike,
+) -> Array:
+    p, q = promote_args_inexact("rel_entr", p, q)
+    both_gt_zero_mask = lax.bitwise_and(lax.gt(p, 0.0), lax.gt(q, 0.0))
+    one_zero_mask = lax.bitwise_and(lax.eq(p, 0.0), lax.ge(q, 0.0))
```
Comment (jakevdp) [inline jax/_src/scipy/special.py:225]: same dtype mismatch issue here.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by SalamanderXing** [comment]
> Thanks!

Thank you @jakevdp for patiently guiding me into this 👍
